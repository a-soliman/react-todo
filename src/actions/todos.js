import uuid from 'uuid';

export const createTodo = 
    (
        {
            title = '', 
            note = '', 
            createdAt = 0
        } = {}
    ) => ({
        type: 'CREATE_TODO',
        todo: {
            id: uuid(),
            title,
            note,
            createdAt
        }
    }
);


export const editTodo = (id, updates) => ({
    type: 'EDIT_TODO',
    id,
    updates
});


export const removeTodo = (id) => ({
    type: 'REMOVE_TODO',
    id
});