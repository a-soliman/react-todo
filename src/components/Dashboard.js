import React from 'react';
import TodosList from './TodosList';
import TodoListFilters from './TodoListFilters';

const DashboardPage = () => (
    <div>
        Todo Dashboard.
        <TodoListFilters />
        <TodosList />
    </div>
);

export default DashboardPage;