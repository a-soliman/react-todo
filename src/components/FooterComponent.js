import React from 'react';

const FooterComponent = () => (
    <div>
        <div className='container fixed-bottom'>
            <div className='row'>
                <div className='col text-center'>Footer</div>
            </div>
        </div>
    </div>
);

export default FooterComponent;