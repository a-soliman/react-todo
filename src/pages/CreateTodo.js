import React from 'react';

const CreateTodoPage = () => (
    <div>
    CreateTodoPage
    </div>
);

export default CreateTodoPage;