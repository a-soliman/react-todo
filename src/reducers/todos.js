const todosReducerDefaultState = [];

const todosReducer =( state = todosReducerDefaultState, action ) => {
    switch (action.type) {
        case 'CREATE_TODO':
            return [...state, action.todo]
        
        case 'REMOVE_TODO':
            return state.filter(todo => todo.id !== action.id)

        case 'EDIT_TODO':
            return state.map(todo => {
                if ( todo.id === action.id ) {
                    return {
                        ...todo,
                        ...action.updates
                    }
                } else {
                    return todo;
                }
            })

        default:
            return state;
    }
};

export default todosReducer;