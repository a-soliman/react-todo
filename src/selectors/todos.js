const gerVisibleTodos = (todos, { text, startDate, endDate }) => {
    return todos.filter( todo => {
        const textMatch = todo.title.toLowerCase().includes(text.toLowerCase()) || todo.note.toLowerCase().includes(text.toLowerCase());
        const startDateMatch = typeof startDate !== 'number' || todo.createdAt >= startDate;
        const endDateMatch = typeof endDate !== 'number' || todo.createdAt <= endDate;

        return textMatch && startDateMatch && endDateMatch;
    });
};

export default gerVisibleTodos;