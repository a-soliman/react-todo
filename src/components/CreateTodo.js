import React from 'react';
import { connect } from 'react-redux';
import TodoForm from './TodoForm';
import { createTodo } from '../actions/todos';

export class CreateTodoPage extends React.Component {
    onSubmit = (todo) => {
        this.props.onSubmit(todo);
        this.props.history.push('/');
    };
    
    render() {
        return (
            <div>
                CreateTodoPage
                <TodoForm  onSubmit={this.onSubmit}/>
            </div>
        );
    }
}

const mapDispatchToProps = (dispatch) => ({
        onSubmit: (todo) => dispatch(createTodo(todo))
});

export default connect(undefined, mapDispatchToProps)(CreateTodoPage);