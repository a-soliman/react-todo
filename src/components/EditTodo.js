import React from 'react';
import { connect } from 'react-redux';
import TodoForm from './TodoForm';
import { editTodo, removeTodo } from '../actions/todos';

const EditTodoPage = (props) => {
    const todo = props.todo;
    
    const onRemoveTodo = () => {
        props.dispatch(removeTodo(todo.id));
        props.history.push('/');
    }

    return (
        <div>
            <TodoForm 
                todo={todo} 
                onSubmit={(updates) => {
                props.dispatch(editTodo(todo.id, updates));
                props.history.push('/');
            }}/>

            <button onClick={onRemoveTodo}>remove</button>
        </div>
    );
}

const mapStateToProps = (state, props) => ({
    todo: state.todos.find(todo => todo.id == props.match.params.id)
});

export default connect(mapStateToProps)(EditTodoPage);