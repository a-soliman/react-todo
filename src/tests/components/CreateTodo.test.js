import React from 'react';
import { shallow } from 'enzyme';
import { CreateTodoPage } from '../../components/CreateTodo';
import { todos } from '../fixtures/todos';

describe('CreateTodo Component', () => {
    test('Should Render CreateTodo Component correctly', () => {
        const wrapper = shallow(<CreateTodoPage />);
        expect(wrapper).toMatchSnapshot();
    });

    test('Should submit with valid form', () => {
        const todo = todos[1];
        const onSubmitSpy = jest.fn();
        const history = { push: jest.fn() };
        const wrapper = shallow(<CreateTodoPage onSubmit={onSubmitSpy} todo={todo} history={history}/>);
        wrapper.find('TodoForm').prop('onSubmit')(todo);
        expect(onSubmitSpy).toHaveBeenLastCalledWith(todo);
        expect(history.push).toHaveBeenLastCalledWith('/');
    })
})