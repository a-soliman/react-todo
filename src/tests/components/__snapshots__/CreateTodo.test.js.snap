// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateTodo Component Should Render CreateTodo Component correctly 1`] = `
<div>
  CreateTodoPage
  <TodoForm
    onSubmit={[Function]}
  />
</div>
`;
